starknet-replay/src/storage/rpc/rpc_client.rs

```diff
@@ -0,0 +1,429 @@
+};
+use starknet_providers::jsonrpc::HttpTransport;
+use starknet_providers::{JsonRpcClient, Provider, ProviderError};
+use tokio::sync::OnceCell;
+use tracing::trace;
+use url::Url;
+
+use crate::block_number::BlockNumber;
+use crate::contract_address::to_field_element;
+use crate::error::DatabaseError;
+use crate::runner::replay_class_hash::ReplayClassHash;
+use crate::storage::rpc::receipt::convert_receipt;
+use crate::storage::rpc::transaction::convert_transaction;
+
+/// This structure partially implements a Starknet RPC client.
+///
+/// The RPC calls included are those needed to replay transactions.
+/// Clone is not derived because it's not supported by Client.
+pub struct RpcClient {
+    /// The endpoint of the Starknet RPC Node.
+    endpoint: Url,
+
+    /// The chain id variable initialised with the first call to
+    /// [`RpcClient::starknet_get_chain_id`]. This is doable because it's not
+    /// possible to replay blocks from different chains.
+    chain_id: OnceCell<ChainId>,
+
+    /// The most recent block number available from the rpc server. It is
+    /// initialised in [`RpcClient::starknet_block_number`]. The assumption
+    /// is that the latest blocks generated after the replay is started are
+    /// ignored.
+    block_number: OnceCell<BlockNumber>,
+}
+impl RpcClient {
+    /// Constructs a new `RpcStorage`.
+    ///
+    /// # Arguments
+    ///
+    /// - `endpoint`: The Url of the Starknet RPC node.
+    #[must_use]
+    pub fn new(endpoint: Url) -> Self {
+        RpcClient {
+            endpoint,
+            chain_id: OnceCell::new(),
+            block_number: OnceCell::new(),
+        }
+    }
+
+    /// This function generates a new client to perform an RPC request to the
+    /// endpoint.
+    ///
+    /// The client can't be shared across threads.
+    fn get_new_client(&self) -> JsonRpcClient<HttpTransport> {
+        JsonRpcClient::new(HttpTransport::new(self.endpoint.clone()))
+    }
+
+    /// This function queries the number of the most recent Starknet block.
+    ///
+    /// # Errors
+    ///
+    /// Returns [`Err`] if the request fails.
+    #[allow(clippy::missing_panics_doc)] // Needed because `tokio::main` calls `unwrap()`
+    #[tokio::main]
+    pub async fn starknet_block_number(&self) -> Result<BlockNumber, DatabaseError> {
+        let block_number: Result<&BlockNumber, DatabaseError> = self
+            .block_number
+            .get_or_try_init(|| async {
+                let block_number: u64 = self.get_new_client().block_number().await?;
+                Ok(BlockNumber::new(block_number))
+            })
+            .await;
+        Ok(block_number?).cloned()
+    }
+
+    /// This function queries the contract class at a specific block.
+    ///
+    /// # Arguments
+    ///
+    /// - `class_hash_at_block`: class hash of the contract to be returned.
+    ///
+    /// # Errors
+    ///
+    /// Returns [`Err`] if the request fails or the class hash doesn't exist.
```

Comment:
Does it return a specific variant of `DatabaseError`? Or multiple variants under different conditions?